### PR TITLE
Kuarora/master

### PR DESCRIPTION
**Pull request by kuarora9091**
enabling stop prompt in platform sdk

### CODE REVIEW

**Comment by saldana** [comment]
### :white_check_mark: Validation status: passed


File | Status | Preview URL | Details
---- | ------ | ----------- | -------
[Skype/Trusted-Application-API/SDK/ClientModel/Resources/AudioVideoFlow.cs](https://github.com/OfficeDev/skype-docs/blob/kuarora/master/Skype/Trusted-Application-API/SDK/ClientModel/Resources/AudioVideoFlow.cs) | :white_check_mark:Succeeded | |

For more details, please refer to the [build report](https://opbuildstorageprod.blob.core.windows.net/report/2017%5C3%5C24%5C5315e434-bbfc-97f8-0e56-7036cb0b5e04%5CPullRequest%5C201703242114384175-242%5Cworkflow_report.html).

**Note:** If you changed an existing file name or deleted a file, broken links in other files to the deleted or renamed file are listed only in the full build report.

File: Skype/Trusted-Application-API/SDK/ClientModel/Resources/AudioVideoFlow.cs [hunk 1]
```diff
@@ -91,6 +91,28 @@ public FlowState State
             return await tcs.Task.ConfigureAwait(false);
         }
 
+        public async Task StopPromptsAsync(LoggingContext loggingContext)
```
Comment (sankum-msft) [inline Skype/Trusted-Application-API/SDK/ClientModel/Resources/AudioVideoFlow.cs]: Following things are also needed to make it usable:
1. Add this method to IAudioVideoFlow interface
2. Add a corresponding capability in AudioVideoCapabilities
3. Update Supports method in AudioVideoFlow class to handle the new capability
4. Add unit tests

File: Skype/Trusted-Application-API/SDK/ClientModel/Resources/AudioVideoFlow.cs [hunk 2]
```diff
@@ -91,6 +91,28 @@ public FlowState State
...
+            {
+                throw new CapabilityNotAvailableException("Link to stop prompts is not available.");
+            }
+
+            var stopPromptLink = UriHelper.CreateAbsoluteUri(BaseUri, href);
+
+            TaskCompletionSource<Prompt> tcs = new TaskCompletionSource<Prompt>();
+            var response = await PostRelatedPlatformResourceAsync(stopPromptLink, null, loggingContext).ConfigureAwait(false);
+
+            if (response != null && response.Headers != null && response.Headers.Location != null)
```
Comment (sankum-msft) [inline Skype/Trusted-Application-API/SDK/ClientModel/Resources/AudioVideoFlow.cs]: This block doesn't serve any purpose for "DeletePrompts"

File: Skype/Trusted-Application-API/SDK/ClientModel/Resources/AudioVideoFlow.cs [hunk 3]
```diff
@@ -91,6 +91,28 @@ public FlowState State
...
+            string href = PlatformResource?.StopPromptsLink?.Href;
+            if (string.IsNullOrWhiteSpace(href))
+            {
+                throw new CapabilityNotAvailableException("Link to stop prompts is not available.");
+            }
+
+            var stopPromptLink = UriHelper.CreateAbsoluteUri(BaseUri, href);
+
+            TaskCompletionSource<Prompt> tcs = new TaskCompletionSource<Prompt>();
+            var response = await PostRelatedPlatformResourceAsync(stopPromptLink, null, loggingContext).ConfigureAwait(false);
```
Comment (sankum-msft) [inline Skype/Trusted-Application-API/SDK/ClientModel/Resources/AudioVideoFlow.cs]: When does this HTTP request return? When all the prompts are signaled to be cancelled or when they have actually been canceled/removed?

File: Skype/Trusted-Application-API/SDK/ClientModel/Resources/AudioVideoFlow.cs (hunk 1 above)
Comment (kuarora9091) [inline Skype/Trusted-Application-API/SDK/ClientModel/Resources/AudioVideoFlow.cs]: done

**Comment by saldana** [comment]
### :x: Validation status: errors


File | Status | Preview URL | Details
---- | ------ | ----------- | -------
[]() | :x:Error | | [Details](#user-content-)

<a name=""></a>
### []()
 - **[Error]** Cannot sync git repo to specified state. Please make sure the pull request is mergeable. Details: Git merge results in conflict, repository remote url: https://github.com/OfficeDev/skype-docs, working directory: W:\1nes-s, target branch: master, target branch commit: 92831beafe4f58ad520dc0fe8c29156af825829b, source branch: kuarora/master, source branch commit: .

For more details, please refer to the [build report](https://opbuildstorageprod.blob.core.windows.net/report/2017%5C3%5C27%5C5315e434-bbfc-97f8-0e56-7036cb0b5e04%5CPullRequest%5C201703271936375416-242%5Cworkflow_report.html).

**Note:** If you changed an existing file name or deleted a file, broken links in other files to the deleted or renamed file are listed only in the full build report.

**Comment by saldana** [comment]
### :x: Validation status: errors


File | Status | Preview URL | Details
---- | ------ | ----------- | -------
[]() | :x:Error | | [Details](#user-content-)

<a name=""></a>
### []()
 - **[Error]** Cannot sync git repo to specified state. Please make sure the pull request is mergeable. Details: Git merge results in conflict, repository remote url: https://github.com/OfficeDev/skype-docs, working directory: W:\lez1-s, target branch: master, target branch commit: 92831beafe4f58ad520dc0fe8c29156af825829b, source branch: kuarora/master, source branch commit: .

For more details, please refer to the [build report](https://opbuildstorageprod.blob.core.windows.net/report/2017%5C3%5C28%5C5315e434-bbfc-97f8-0e56-7036cb0b5e04%5CPullRequest%5C201703282240152060-242%5Cworkflow_report.html).

**Note:** If you changed an existing file name or deleted a file, broken links in other files to the deleted or renamed file are listed only in the full build report.

**Comment by saldana** [comment]
### :white_check_mark: Validation status: passed


File | Status | Preview URL | Details
---- | ------ | ----------- | -------
[Skype/Trusted-Application-API/SDK/ClientModel/Capabilities/Capabilities.cs](https://github.com/OfficeDev/skype-docs/blob/kuarora/master/Skype/Trusted-Application-API/SDK/ClientModel/Capabilities/Capabilities.cs) | :white_check_mark:Succeeded | |
[Skype/Trusted-Application-API/SDK/ClientModel/Interfaces/Resources.cs](https://github.com/OfficeDev/skype-docs/blob/kuarora/master/Skype/Trusted-Application-API/SDK/ClientModel/Interfaces/Resources.cs) | :white_check_mark:Succeeded | |
[Skype/Trusted-Application-API/SDK/ClientModel/Resources/AudioVideoFlow.cs](https://github.com/OfficeDev/skype-docs/blob/kuarora/master/Skype/Trusted-Application-API/SDK/ClientModel/Resources/AudioVideoFlow.cs) | :white_check_mark:Succeeded | |
[Skype/Trusted-Application-API/SDK/Tests/ClientModel/AudioVideoFlow.cs](https://github.com/OfficeDev/skype-docs/blob/kuarora/master/Skype/Trusted-Application-API/SDK/Tests/ClientModel/AudioVideoFlow.cs) | :white_check_mark:Succeeded | |
[Skype/Trusted-Application-API/SDK/Tests/Data/Event_AudioVideoFlowConnected.json](https://github.com/OfficeDev/skype-docs/blob/kuarora/master/Skype/Trusted-Application-API/SDK/Tests/Data/Event_AudioVideoFlowConnected.json) | :white_check_mark:Succeeded | |
[Skype/Trusted-Application-API/SDK/Tests/Data/MockResponseData.json](https://github.com/OfficeDev/skype-docs/blob/kuarora/master/Skype/Trusted-Application-API/SDK/Tests/Data/MockResponseData.json) | :white_check_mark:Succeeded | |
[Skype/Trusted-Application-API/SDK/Tests/DataUrls.cs](https://github.com/OfficeDev/skype-docs/blob/kuarora/master/Skype/Trusted-Application-API/SDK/Tests/DataUrls.cs) | :white_check_mark:Succeeded | |

For more details, please refer to the [build report](https://opbuildstorageprod.blob.core.windows.net/report/2017%5C3%5C28%5C5315e434-bbfc-97f8-0e56-7036cb0b5e04%5CPullRequest%5C201703282300242363-242%5Cworkflow_report.html).

**Note:** If you changed an existing file name or deleted a file, broken links in other files to the deleted or renamed file are listed only in the full build report.

File: Skype/Trusted-Application-API/SDK/ClientModel/Resources/AudioVideoFlow.cs [hunk 4]
```diff
@@ -91,6 +94,40 @@ public FlowState State
...
+            await PostRelatedPlatformResourceAsync(stopPromptLink, null, loggingContext).ConfigureAwait(false);
+
+            foreach (KeyValuePair<string, TaskCompletionSource<Prompt>> entry in m_onGoingPromptTcses)
+            {
+                try
+                {
+                    await entry.Value.Task.ConfigureAwait(false);
+                } catch (RemotePlatformServiceException psException)
+                {
+                    if (psException.ErrorInformation.Subcode == ResourceModel.ErrorSubcode.PromptsStopped)
```
Comment (sankum-msft) [inline Skype/Trusted-Application-API/SDK/ClientModel/Resources/AudioVideoFlow.cs]: check psException.ErrorInformation.Code also

File: Skype/Trusted-Application-API/SDK/ClientModel/Resources/AudioVideoFlow.cs [hunk 5]
```diff
@@ -91,6 +94,40 @@ public FlowState State
...
+                    {
+                        Logger.Instance.Information("Exception in waiting for media prompt to stop");
```
Comment (sankum-msft) [inline Skype/Trusted-Application-API/SDK/ClientModel/Resources/AudioVideoFlow.cs]: "Prompt stopped"; do not mention exception in the log line

File: Skype/Trusted-Application-API/SDK/Tests/ClientModel/AudioVideoFlow.cs [hunk 6]
```diff
@@ -197,6 +211,33 @@ public void StatePropertyShouldGetUpdatedOnEvent()
...
+            m_restfulClient.HandleRequestProcessed += (sender, args) =>
+            {
+                if (args.Uri == new Uri(DataUrls.StopPrompts) && args.Method == HttpMethod.Post)
+                {
+                    requestReceived.SetResult(true);
+                }
+            };
+
+            // When
+            Task promptTask = m_audioVideoFlow.PlayPromptAsync(new Uri("https://example.com/prompt"), m_loggingContext);
```
Comment (sankum-msft) [inline Skype/Trusted-Application-API/SDK/Tests/ClientModel/AudioVideoFlow.cs]: you want to wait for the Task returned by StopPromptAsync(); not by PlayPromptAsync(). I mean promptTask should be assigned the return value of StopPromptAsync()

File: Skype/Trusted-Application-API/SDK/Tests/ClientModel/AudioVideoFlow.cs [hunk 7]
```diff
@@ -217,6 +258,28 @@ public void StatePropertyShouldGetUpdatedOnEvent()
             await requestReceived.Task.TimeoutAfterAsync(TimeSpan.FromMilliseconds(200)).ConfigureAwait(false);
         }
 
+
+        [TestMethod]
+        public async Task StopPromptsShouldWorkWithNullLoggingContext()
```
Comment (sankum-msft) [inline Skype/Trusted-Application-API/SDK/Tests/ClientModel/AudioVideoFlow.cs]: This case would make more sense/test more code if a prompt was running

File: Skype/Trusted-Application-API/SDK/Tests/ClientModel/AudioVideoFlow.cs [hunk 8]
```diff
@@ -197,6 +211,33 @@ public void StatePropertyShouldGetUpdatedOnEvent()
         }
 
         [TestMethod]
+        public async Task StopPromptsShouldReturnOnlyOnPromptCompletedEvent()
```
Comment (sankum-msft) [inline Skype/Trusted-Application-API/SDK/Tests/ClientModel/AudioVideoFlow.cs]: Add following test cases as well:
1. StopPromptAsync called when no prompt is being played
2. StopPromptAsync called when multiple prompts are in the playing queue

File: Skype/Trusted-Application-API/SDK/ClientModel/Resources/AudioVideoFlow.cs (hunk 4 above)
Comment (kuarora9091) [inline Skype/Trusted-Application-API/SDK/ClientModel/Resources/AudioVideoFlow.cs]: do we need as it doesn't matter?

File: Skype/Trusted-Application-API/SDK/ClientModel/Resources/AudioVideoFlow.cs (hunk 5 above)
Comment (kuarora9091) [inline Skype/Trusted-Application-API/SDK/ClientModel/Resources/AudioVideoFlow.cs]: removed.

File: Skype/Trusted-Application-API/SDK/Tests/ClientModel/AudioVideoFlow.cs (hunk 8 above)
Comment (kuarora9091) [inline Skype/Trusted-Application-API/SDK/Tests/ClientModel/AudioVideoFlow.cs]: I thought of first one but it really doesn't matter. I will do both.

Comment (kuarora9091) [inline Skype/Trusted-Application-API/SDK/Tests/ClientModel/AudioVideoFlow.cs]: case for no play prompt is already handled in null logging context

**Comment by saldana** [comment]
### :white_check_mark: Validation status: passed


File | Status | Preview URL | Details
---- | ------ | ----------- | -------
[Skype/Trusted-Application-API/SDK/ClientModel/Capabilities/Capabilities.cs](https://github.com/OfficeDev/skype-docs/blob/kuarora/master/Skype/Trusted-Application-API/SDK/ClientModel/Capabilities/Capabilities.cs) | :white_check_mark:Succeeded | |
[Skype/Trusted-Application-API/SDK/ClientModel/Interfaces/Resources.cs](https://github.com/OfficeDev/skype-docs/blob/kuarora/master/Skype/Trusted-Application-API/SDK/ClientModel/Interfaces/Resources.cs) | :white_check_mark:Succeeded | |
[Skype/Trusted-Application-API/SDK/ClientModel/Resources/AudioVideoFlow.cs](https://github.com/OfficeDev/skype-docs/blob/kuarora/master/Skype/Trusted-Application-API/SDK/ClientModel/Resources/AudioVideoFlow.cs) | :white_check_mark:Succeeded | |
[Skype/Trusted-Application-API/SDK/Tests/ClientModel/AudioVideoFlow.cs](https://github.com/OfficeDev/skype-docs/blob/kuarora/master/Skype/Trusted-Application-API/SDK/Tests/ClientModel/AudioVideoFlow.cs) | :white_check_mark:Succeeded | |
[Skype/Trusted-Application-API/SDK/Tests/Data/Event_AudioVideoFlowConnected.json](https://github.com/OfficeDev/skype-docs/blob/kuarora/master/Skype/Trusted-Application-API/SDK/Tests/Data/Event_AudioVideoFlowConnected.json) | :white_check_mark:Succeeded | |
[Skype/Trusted-Application-API/SDK/Tests/Data/MockResponseData.json](https://github.com/OfficeDev/skype-docs/blob/kuarora/master/Skype/Trusted-Application-API/SDK/Tests/Data/MockResponseData.json) | :white_check_mark:Succeeded | |
[Skype/Trusted-Application-API/SDK/Tests/DataUrls.cs](https://github.com/OfficeDev/skype-docs/blob/kuarora/master/Skype/Trusted-Application-API/SDK/Tests/DataUrls.cs) | :white_check_mark:Succeeded | |

For more details, please refer to the [build report](https://opbuildstorageprod.blob.core.windows.net/report/2017%5C3%5C29%5C5315e434-bbfc-97f8-0e56-7036cb0b5e04%5CPullRequest%5C201703290001225900-242%5Cworkflow_report.html).

**Note:** If you changed an existing file name or deleted a file, broken links in other files to the deleted or renamed file are listed only in the full build report.

**Comment by saldana** [comment]
### :white_check_mark: Validation status: passed


File | Status | Preview URL | Details
---- | ------ | ----------- | -------
[Skype/Trusted-Application-API/SDK/ClientModel/Capabilities/Capabilities.cs](https://github.com/OfficeDev/skype-docs/blob/kuarora/master/Skype/Trusted-Application-API/SDK/ClientModel/Capabilities/Capabilities.cs) | :white_check_mark:Succeeded | |
[Skype/Trusted-Application-API/SDK/ClientModel/Interfaces/Resources.cs](https://github.com/OfficeDev/skype-docs/blob/kuarora/master/Skype/Trusted-Application-API/SDK/ClientModel/Interfaces/Resources.cs) | :white_check_mark:Succeeded | |
[Skype/Trusted-Application-API/SDK/ClientModel/Resources/AudioVideoFlow.cs](https://github.com/OfficeDev/skype-docs/blob/kuarora/master/Skype/Trusted-Application-API/SDK/ClientModel/Resources/AudioVideoFlow.cs) | :white_check_mark:Succeeded | |
[Skype/Trusted-Application-API/SDK/Tests/ClientModel/AudioVideoFlow.cs](https://github.com/OfficeDev/skype-docs/blob/kuarora/master/Skype/Trusted-Application-API/SDK/Tests/ClientModel/AudioVideoFlow.cs) | :white_check_mark:Succeeded | |
[Skype/Trusted-Application-API/SDK/Tests/Data/Event_AudioVideoFlowConnected.json](https://github.com/OfficeDev/skype-docs/blob/kuarora/master/Skype/Trusted-Application-API/SDK/Tests/Data/Event_AudioVideoFlowConnected.json) | :white_check_mark:Succeeded | |
[Skype/Trusted-Application-API/SDK/Tests/Data/Event_Prompt2Started.json](https://github.com/OfficeDev/skype-docs/blob/kuarora/master/Skype/Trusted-Application-API/SDK/Tests/Data/Event_Prompt2Started.json) | :white_check_mark:Succeeded | |
[Skype/Trusted-Application-API/SDK/Tests/Data/Event_Prompt2Stopped.json](https://github.com/OfficeDev/skype-docs/blob/kuarora/master/Skype/Trusted-Application-API/SDK/Tests/Data/Event_Prompt2Stopped.json) | :white_check_mark:Succeeded | |
[Skype/Trusted-Application-API/SDK/Tests/Data/Event_PromptStopped.json](https://github.com/OfficeDev/skype-docs/blob/kuarora/master/Skype/Trusted-Application-API/SDK/Tests/Data/Event_PromptStopped.json) | :white_check_mark:Succeeded | |
[Skype/Trusted-Application-API/SDK/Tests/Data/MockResponseData.json](https://github.com/OfficeDev/skype-docs/blob/kuarora/master/Skype/Trusted-Application-API/SDK/Tests/Data/MockResponseData.json) | :white_check_mark:Succeeded | |
[Skype/Trusted-Application-API/SDK/Tests/DataUrls.cs](https://github.com/OfficeDev/skype-docs/blob/kuarora/master/Skype/Trusted-Application-API/SDK/Tests/DataUrls.cs) | :white_check_mark:Succeeded | |
[Skype/Trusted-Application-API/SDK/Tests/Mocks/MockRestfulClient.cs](https://github.com/OfficeDev/skype-docs/blob/kuarora/master/Skype/Trusted-Application-API/SDK/Tests/Mocks/MockRestfulClient.cs) | :white_check_mark:Succeeded | |

For more details, please refer to the [build report](https://opbuildstorageprod.blob.core.windows.net/report/2017%5C3%5C29%5C5315e434-bbfc-97f8-0e56-7036cb0b5e04%5CPullRequest%5C201703292210564595-242%5Cworkflow_report.html).

**Note:** If you changed an existing file name or deleted a file, broken links in other files to the deleted or renamed file are listed only in the full build report.

**Pull request merged**